From e6a2a79fd296cc89e9f687406626ca6050e75c66 Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Sat, 24 Jun 2017 12:10:19 -0400
Subject: [PATCH] CPU exhaustion in ReadOneDJVUImag

Due to lack of format validation, a crafted file will cause a loop to run endless.

bug: https://github.com/ImageMagick/ImageMagick/issues/528
bug-debian: https://bugs.debian.org/867826
origin: https://github.com/ImageMagick/ImageMagick/commit/78b819628b6a9429f0c33b72e695b4df0b32faea

(cherry picked from commit 78b819628b6a9429f0c33b72e695b4df0b32faea)
---
 coders/djvu.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/coders/djvu.c b/coders/djvu.c
index 973e8ff85..42a3e7e68 100644
--- a/coders/djvu.c
+++ b/coders/djvu.c
@@ -613,6 +613,7 @@ static Image *ReadOneDJVUImage(LoadContext* lc,const int pagenum,
                                 if (tag == 0) break;
                                 ddjvu_message_pop(lc->context);
                         } while ((message = ddjvu_message_peek(lc->context)));
+                if (tag == 0) break;
         } while (!ddjvu_page_decoding_done(lc->page));
 
         ddjvu_document_get_pageinfo(lc->document, pagenum, &info);
@@ -887,7 +888,8 @@ static Image *ReadDJVUImage(const ImageInfo *image_info,
         break;
   }
   djvu_close_lc(lc);
-  (void) CloseBlob(images);
+  if (images != (Image *) NULL)
+    (void) CloseBlob(images);
   if (image != (Image *) NULL)
     image=DestroyImageList(image);
 
